File: Python3/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    N, M, answer = len(maps), len(maps[0]), 10001 #Map 크기 저장
    queue = deque()
    dr, dc = [-1, 1, 0, 0], [0, 0, -1, 1]
    queue.append([0, 0, 1])
    
    visited = [[False for j in i] for i in maps]
    r, c, cnt = 0, 0, 0
    while queue:
        r, c, cnt = queue.popleft()
        if r == N - 1 and c == M - 1:
            answer = min(answer, cnt)
        for nr, nc in zip(dr, dc):
            nr, nc = nr + r, nc + c
            if 0 <= nr and nr < N and 0 <= nc and nc < M: #maps 범위 내일 경우
                if not visited[nr][nc] and maps[nr][nc] == 1:
                    visited[nr][nc] = True
                    queue.append([nr, nc, cnt + 1])
    return answer if answer != 10001 else -1
    
        
            
    

File: C++/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>
using namespace std;

int solution(vector<vector<int> > maps) {
    int answer = 10001, nr, nc, N = maps.size(), M = maps[0].size();
    queue<vector<int>> q;
    q.push({0, 0, 1});
    vector<int> dr = {-1, 1, 0, 0}, dc = {0, 0, -1, 1};
    vector<vector<bool>> visited(N);
    for(int i = 0; i < N; i++) {
        visited[i] = vector<bool>(M);
    }
    
    vector<int> node;
    while(!q.empty()) {
        node = q.front(); q.pop();
        if(node[0] == N - 1 && node[1] == M - 1) {
            answer = min(answer, node[2]);
        }
        
        for(int i = 0; i < 4; i++) {
            nr = node[0] + dr[i];
            nc = node[1] + dc[i];
            if(0 <= nr && nr < N && 0 <= nc && nc < M && !visited[nr][nc] && maps[nr][nc] == 1) {
                visited[nr][nc] = true;
                q.push({nr, nc, node[2] + 1});
            }
        }
    }
    return answer == 10001 ? -1:answer;
}